Source/Dafny/Util.cs

```diff
@@ -817,13 +818,24 @@ public class ExpressionTester {
 
       if (expr is IdentifierExpr expression) {
         if (expression.Var != null && expression.Var.IsGhost) {
-          reporter?.Error(MessageSource.Resolver, expression, "ghost variables are allowed only in specification contexts");
+          reporter?.Error(MessageSource.Resolver, expression, "a ghost variable is allowed only in specification contexts");
           return false;
         }
 
       } else if (expr is MemberSelectExpr selectExpr) {
         if (selectExpr.Member != null && selectExpr.Member.IsGhost) {
-          reporter?.Error(MessageSource.Resolver, selectExpr, "ghost fields are allowed only in specification contexts");
+          string what;
+          switch (selectExpr.Member) {
```

Comment:
Could you please create a new field for all members, named like `WhatKindWithExplicitGhostness` or something similar and preferably shorter, that will return what you compute in the variable "what"?
I bet this will be useful for reporting other errors, so it might be worth it moving it there..